### ISSUE TITLE
aead: generate_nonce accept an immutable rng

### ISSUE THREAD

**Issue by alxiong** [RustCrypto/traits#1087]
I'm a little confused by the latest `AeadCore::generate_nonce()` API:

in [code](https://github.com/RustCrypto/traits/blob/aead-v0.5.1/aead/src/lib.rs), it seems to be 
```rust
    #[cfg(feature = "rand_core")]
    #[cfg_attr(docsrs, doc(cfg(feature = "rand_core")))]
    fn generate_nonce(mut rng: impl CryptoRng + RngCore) -> Nonce<Self>
    where
        Nonce<Self>: Default,
    {
        let mut nonce = Nonce::<Self>::default();
        rng.fill_bytes(&mut nonce);
        nonce
    }
```

however in [rustdoc](https://docs.rs/aead/latest/aead/trait.AeadCore.html#method.generate_nonce) (latest/v0.5.1), it seem to be accepting an immutable `rng`: `fn generate_nonce(rng: impl CryptoRng + RngCore) -> Nonce<Self>`

that's the first confusion.

The second confusion is, why does it accept `mut rng` instead of `&mut rng`?

**Comment by tarcieri** [comment]
The `mut` keyword is a property of the `rng` local binding.

It's leveraging a blanket impl of `RngCore` for `&mut R`: https://docs.rs/rand_core/latest/rand_core/trait.RngCore.html#impl-RngCore-for-%26%27a%20mut%20R

The purpose of using this was to make the type signature a bit cleaner than `rng: &mut (impl CryptoRng + RngCore)`.

@newpavlov it seems people get really confused by these type signatures because they don't immediately realize they need to pass `&mut R`. See also:

https://github.com/RustCrypto/traits/pull/1062

**Comment by newpavlov** [comment]
In addition to cleaner signatures the current approach is also allows us to write `Alg::generate_nonce(OsRng)` and `Alg::generate_nonce(thread_rng())` without redundant `&mut`s.

I am open to changing it in the next breaking releases, but I think we should use `&mut R` and `where R: CryptoRng + RngCore`. `&mut impl Trait` looks a bit weird in my opinion.

**Comment by tarcieri** [comment]
Being able to directly pass `OsRng` is cool. I didn’t even think of that.

I’d be down to change it back to a named parameter in the next release, though. I think it’s just too confusing as is.

**Comment by alxiong** [comment]
I see, I was so entrenched in seeing and using function signature like `&mut R`, now I actually king of like the current API design. 

Future reference for myself:

```rust
trait RngCore {}

#[derive(Default)]
struct OsRng;
impl RngCore for OsRng {}

#[derive(Default)]
struct ChaChaRng;
impl RngCore for ChaChaRng {}

// blanket implementation, mimicking rand_core: https://docs.rs/rand_core/latest/rand_core/trait.RngCore.html#impl-RngCore-for-%26%27a%20mut%20R
impl<'a, R: RngCore> RngCore for &'a mut R {}

// same as, but generic over R looks better:
// fn encrypt(_rng: &mut impl RngCore) {}
// passing in an immutable binding of mutable reference: can't change what `rng` points to, can change the value/state that `rng` points to.
fn encrypt<R: RngCore>(_rng: &mut R) {}

// what RustCrypto and Halo2 uses. Cleaner-looking API than `encrypt()`
fn encrypt2(mut _rng: impl RngCore) {}

fn main() {
    let mut rng = OsRng::default();

    // encrypt(rng); // will give clear error:  "expected `&mut _`, consider mutably borrowing here: `&mut rng`"
    encrypt(&mut rng); // only this will works! obvious from function signature

    // encrypt2(&rng); // the trait `RngCore` is not implemented for `&OsRng`
    //                 // will be troublesome if we can pass in `&rng`, because that means inside `encrypt()`, they can mutate the reference and point
    //                 // to some other Rng, which is obviously undesirable. Thankfully, the blanket implementation is only for `&mut R`, not `&R`, not `&R`

    encrypt2(&mut rng); // works! becasue of the blanket impl
    encrypt2(rng); // works! but note that we move `rng` into this function, useful when you only want to use some Rng once without the redundant &mut
}
```

**Comment by tarcieri** [comment]
Something else I think is a bit annoying is the need to declare the binding `mut` and reborrow if you already have a `&mut Rng`, e.g.

```rust
pub fn some_fn(mut rng: &mut Rng) -> Nonce {
    MyAead::generate_nonce(&mut rng)
}
```

...whereas with `<R: RngCore>(rng: &mut R)` it's this:


```rust
pub fn some_fn(rng: &mut Rng) -> Nonce {
    MyAead::generate_nonce(rng)
}
```

**Comment by alxiong** [comment]
yeah, that's true. 

Guess no clear winner between the two options then, and a matter of taste.

**Comment by alxiong** [comment]
another tangential question, why are we marking `rng: RngCore + CryptoRng` instead of `rng: Rng + CryptoRng` in user-facing APIs? @tarcieri 

didn't [`rng_core`'s doc](https://docs.rs/rand/latest/rand/trait.RngCore.html) suggests that only Rng implementations should impl `RngCore`, otherwise "users should normally use `Rng` trait, which is automatically implemented for every type implementing `RngCore`."?

**Comment by tarcieri** [comment]
TIL: `rand_core` added a `CryptoRngCore` core extension trait in the latest release (v0.6.4):

https://docs.rs/rand_core/latest/rand_core/trait.CryptoRngCore.html

Given that, I think we could move to `&mut impl CryptoRngCore` in the next breaking release.

(though for whatever reason `rand_core` docs recommend `&mut dyn CryptoRngCore`)

**Comment by tarcieri** [comment]
I think this issue is "solved", and opened #1148 for further discussion.
